src/query/storage/m3/storage.go

```diff
@@ -257,20 +260,28 @@ func (s *m3storage) Write(
 	id := query.Tags.ID()
 	// TODO: Consider caching id -> identID
 	identID := ident.StringID(id)
-	common := &writeRequestCommon{
-		store:       s,
-		annotation:  query.Annotation,
-		unit:        query.Unit,
-		id:          identID,
-		tagIterator: storage.TagsToIdentTagIterator(query.Tags),
-		attributes:  query.Attributes,
-	}
+	tagIterator := storage.TagsToIdentTagIterator(query.Tags)
+
+	var (
```

Comment:
Considering how this is being broken up, it may be a good idea to use the pooled workerpools like read does? That way every incoming request will get a set of workers for all datapoints in it and we don't have situations where some requests get starved 

If we want to simplify it a bit, can revisit my m3x PR at: https://github.com/m3db/m3x/pull/181 and make it a first class citizen, then refactor the readerpools too?

Alternatively, just use the same pool as the readerpools here if you agree with taking this approach

Reply:
Can you explain the rationale for pooled worker pools? Kind of confused by that

Reply:
General idea was to give each incoming request its own pool while having a static max size, so that the more expensive requests would not tie up smaller requests, if that's not the correct approach, happy to drop it and revert to your pooled worker pool

Reply:
Sync'd up in person and settled on (pending discussion with @nikunjgit) replacing the existing WorkerPool with PooledWorkerPool and then getting rid of the generic ObjectPool on top of that

Reply:
i'll use a PooledWorkerPool for writes and let @arnikola merge the read object pool in another diff. Sound good ?

Reply:
Sure